File: android/app/src/main/java/com/invoiceapp/util/AccessibilityHelper.kt
package com.invoiceapp.util

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.*
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.sp
import java.text.NumberFormat
import java.util.*

/**
 * Accessibility utilities for Invoice App
 */
object AccessibilityHelper {

    /**
     * Format currency for screen readers
     */
    fun formatCurrencyForAccessibility(amount: Double): String {
        val formatter = NumberFormat.getCurrencyInstance(Locale.US)
        val formattedAmount = formatter.format(amount)
        return formattedAmount.replace("$", "").let { "$it dollars" }
    }

    /**
     * Format date for screen readers
     */
    fun formatDateForAccessibility(date: String): String {
        return date.replace("-", " ")
    }

    /**
     * Create content description for invoice item
     */
    fun invoiceItemDescription(
        vendorName: String,
        amount: Double,
        date: String,
        category: String?
    ): String {
        val amountText = formatCurrencyForAccessibility(amount)
        val categoryText = category?.let { ", Category: $it" } ?: ""
        return "Invoice from $vendorName, Amount: $amountText, Date: $date$categoryText"
    }

    /**
     * Create content description for budget item
     */
    fun budgetItemDescription(
        name: String,
        spent: Double,
        total: Double,
        percentage: Double
    ): String {
        val spentText = formatCurrencyForAccessibility(spent)
        val totalText = formatCurrencyForAccessibility(total)
        val status = when {
            percentage >= 100 -> "exceeded"
            percentage >= 80 -> "near limit"
            else -> "on track"
        }
        return "Budget: $name, Spent: $spentText of $totalText, ${percentage.toInt()} percent used, $status"
    }

    /**
     * Create content description for category item
     */
    fun categoryItemDescription(
        name: String,
        description: String?,
        isDefault: Boolean
    ): String {
        val descText = description?.let { ", $it" } ?: ""
        val defaultText = if (isDefault) ", Default category" else ""
        return "Category: $name$descText$defaultText"
    }

    /**
     * Scale text for accessibility
     */
    @Composable
    fun scaledTextStyle(baseStyle: TextStyle, scale: Float = 1f): TextStyle {
        return baseStyle.copy(fontSize = (baseStyle.fontSize.value * scale).sp)
    }

    /**
     * Minimum touch target size (48dp as per Material Design)
     */
    const val MinTouchTargetSize = 48
}

/**
 * Extension function to add accessibility semantics to Modifier
 */
fun Modifier.accessibilityLabel(
    label: String,
    role: Role? = null
): Modifier = this.semantics {
    contentDescription = label
    role?.let { this.role = it }
}

/**
 * Extension function for clickable items with accessibility
 */
@Composable
fun Modifier.accessibleClickable(
    label: String,
    role: Role = Role.Button,
    enabled: Boolean = true,
    onClick: () -> Unit
): Modifier {
    val interactionSource = remember { MutableInteractionSource() }
    return this
        .semantics {
            contentDescription = label
            this.role = role
            if (!enabled) {
                disabled()
            }
        }
        .clickable(
            enabled = enabled,
            onClick = onClick,
            interactionSource = interactionSource,
            indication = null
        )
}

/**
 * Extension function for buttons with loading state
 */
fun Modifier.accessibilityState(
    isLoading: Boolean = false,
    isError: Boolean = false,
    errorMessage: String? = null
): Modifier = this.semantics {
    if (isLoading) {
        stateDescription = "Loading"
    }
    if (isError && errorMessage != null) {
        liveRegion = LiveRegionMode.Assertive
        error(errorMessage)
    }
}

/**
 * Extension function for progress indicators
 */
fun Modifier.accessibilityProgress(
    current: Float,
    total: Float,
    label: String
): Modifier = this.semantics {
    contentDescription = "$label: ${(current / total * 100).toInt()} percent"
    progressBarRangeInfo = ProgressBarRangeInfo(current, 0f..total)
}

/**
 * Extension function for form fields
 */
fun Modifier.accessibilityFormField(
    label: String,
    value: String,
    isError: Boolean = false,
    errorMessage: String? = null
): Modifier = this.semantics {
    contentDescription = "$label: $value"
    if (isError && errorMessage != null) {
        error(errorMessage)
    }
}

/**
 * Extension function for headings
 */
fun Modifier.accessibilityHeading(): Modifier = this.semantics {
    heading()
}

/**
 * Extension function for lists
 */
fun Modifier.accessibilityCollection(
    itemCount: Int,
    collectionType: String = "list"
): Modifier = this.semantics {
    contentDescription = "$collectionType with $itemCount items"
}

/**
 * Extension function for images
 */
fun Modifier.accessibilityImage(
    description: String,
    isDecorative: Boolean = false
): Modifier = this.semantics {
    if (isDecorative) {
        contentDescription = ""
    } else {
        contentDescription = description
        role = Role.Image
    }
}

File: ios/InvoiceApp/InvoiceApp/Utilities/AccessibilityHelper.swift
import SwiftUI

/**
 * Accessibility utilities for Invoice App
 */

// MARK: - Accessibility Constants
enum AccessibilityConstants {
    static let minimumTouchTarget: CGFloat = 44
}

// MARK: - Accessibility Helper
struct AccessibilityHelper {
    /// Format currency for VoiceOver
    static func formatCurrencyForVoiceOver(_ amount: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        let formatted = formatter.string(from: amount as NSDecimalNumber) ?? "$0.00"
        return formatted.replacingOccurrences(of: "$", with: "").appending(" dollars")
    }

    /// Format date for VoiceOver
    static func formatDateForVoiceOver(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateStyle = .long
        formatter.timeStyle = .none
        return formatter.string(from: date)
    }

    /// Create invoice item accessibility label
    static func invoiceItemLabel(
        vendorName: String,
        amount: Decimal,
        date: Date,
        category: String?
    ) -> String {
        let amountText = formatCurrencyForVoiceOver(amount)
        let dateText = formatDateForVoiceOver(date)
        let categoryText = category.map { ", Category: \($0)" } ?? ""
        return "Invoice from \(vendorName), Amount: \(amountText), Date: \(dateText)\(categoryText)"
    }

    /// Create budget item accessibility label
    static func budgetItemLabel(
        name: String,
        spent: Decimal,
        total: Decimal,
        percentage: Double
    ) -> String {
        let spentText = formatCurrencyForVoiceOver(spent)
        let totalText = formatCurrencyForVoiceOver(total)
        let status: String
        if percentage >= 100 {
            status = "exceeded"
        } else if percentage >= 80 {
            status = "near limit"
        } else {
            status = "on track"
        }
        return "Budget: \(name), Spent: \(spentText) of \(totalText), \(Int(percentage)) percent used, \(status)"
    }

    /// Create category item accessibility label
    static func categoryItemLabel(
        name: String,
        description: String?,
        isDefault: Bool
    ) -> String {
        let descText = description.map { ", \($0)" } ?? ""
        let defaultText = isDefault ? ", Default category" : ""
        return "Category: \(name)\(descText)\(defaultText)"
    }

    /// Create button accessibility label with state
    static func buttonLabel(_ text: String, isLoading: Bool = false, isDisabled: Bool = false) -> String {
        if isLoading {
            return "\(text), Loading"
        } else if isDisabled {
            return "\(text), Disabled"
        }
        return text
    }
}

// MARK: - View Modifiers

struct AccessibilityLabelModifier: ViewModifier {
    let label: String
    let hint: String?
    let traits: AccessibilityTraits

    func body(content: Content) -> some View {
        content
            .accessibilityLabel(label)
            .accessibilityHint(hint ?? "")
            .accessibilityAddTraits(traits)
    }
}

struct AccessibilityHeadingModifier: ViewModifier {
    let level: Int

    func body(content: Content) -> some View {
        content
            .accessibilityAddTraits(.isHeader)
            .accessibilityLabel("Heading level \(level)")
    }
}

struct AccessibilityButtonModifier: ViewModifier {
    let label: String
    let hint: String?
    let isDisabled: Bool

    func body(content: Content) -> some View {
        content
            .accessibilityLabel(label)
            .accessibilityHint(hint ?? "")
            .accessibilityAddTraits(.isButton)
            .accessibilityRemoveTraits(isDisabled ? [] : .isButton)
            .accessibilityAddTraits(isDisabled ? .isButton : [])
    }
}

struct AccessibilityImageModifier: ViewModifier {
    let description: String?
    let isDecorative: Bool

    func body(content: Content) -> some View {
        if isDecorative {
            content
                .accessibilityHidden(true)
        } else {
            content
                .accessibilityLabel(description ?? "")
                .accessibilityAddTraits(.isImage)
        }
    }
}

struct MinimumTouchTargetModifier: ViewModifier {
    func body(content: Content) -> some View {
        content
            .frame(minWidth: AccessibilityConstants.minimumTouchTarget,
                   minHeight: AccessibilityConstants.minimumTouchTarget)
    }
}

struct AccessibilityValueModifier: ViewModifier {
    let value: String
    let range: ClosedRange<Double>?
    let current: Double?

    func body(content: Content) -> some View {
        var view = content.accessibilityValue(value)

        if let range = range, let current = current {
            view = view.accessibilityAdjustableAction { direction in
                switch direction {
                case .increment:
                    // Handle increment
                    break
                case .decrement:
                    // Handle decrement
                    break
                @unknown default:
                    break
                }
            }
        }

        return view
    }
}

// MARK: - View Extensions

extension View {
    /// Add accessibility label with optional hint and traits
    func accessibilityLabel(
        _ label: String,
        hint: String? = nil,
        traits: AccessibilityTraits = []
    ) -> some View {
        modifier(AccessibilityLabelModifier(label: label, hint: hint, traits: traits))
    }

    /// Mark view as heading
    func accessibilityHeading(level: Int = 1) -> some View {
        modifier(AccessibilityHeadingModifier(level: level))
    }

    /// Add button accessibility
    func accessibilityButton(
        label: String,
        hint: String? = nil,
        isDisabled: Bool = false
    ) -> some View {
        modifier(AccessibilityButtonModifier(label: label, hint: hint, isDisabled: isDisabled))
    }

    /// Add image accessibility
    func accessibilityImage(
        description: String? = nil,
        isDecorative: Bool = false
    ) -> some View {
        modifier(AccessibilityImageModifier(description: description, isDecorative: isDecorative))
    }

    /// Ensure minimum touch target size
    func minimumTouchTarget() -> some View {
        modifier(MinimumTouchTargetModifier())
    }

    /// Add accessibility value with optional adjustable range
    func accessibilityValue(
        _ value: String,
        range: ClosedRange<Double>? = nil,
        current: Double? = nil
    ) -> some View {
        modifier(AccessibilityValueModifier(value: value, range: range, current: current))
    }

    /// Group accessibility elements
    func accessibilityGroup() -> some View {
        self.accessibilityElement(children: .combine)
    }

    /// Make view ignored by accessibility
    func accessibilityIgnored() -> some View {
        self.accessibilityHidden(true)
    }
}

// MARK: - Accessibility-Friendly Components

/// Accessible button with minimum touch target
struct AccessibleButton<Label: View>: View {
    let action: () -> Void
    let label: String
    let hint: String?
    let isDisabled: Bool
    let content: () -> Label

    init(
        _ label: String,
        hint: String? = nil,
        isDisabled: Bool = false,
        action: @escaping () -> Void,
        @ViewBuilder content: @escaping () -> Label
    ) {
        self.label = label
        self.hint = hint
        self.isDisabled = isDisabled
        self.action = action
        self.content = content
    }

    var body: some View {
        Button(action: action) {
            content()
                .frame(minWidth: AccessibilityConstants.minimumTouchTarget,
                       minHeight: AccessibilityConstants.minimumTouchTarget)
        }
        .disabled(isDisabled)
        .accessibilityLabel(label)
        .accessibilityHint(hint ?? "")
        .accessibilityAddTraits(.isButton)
    }
}

/// Accessible text field with label
struct AccessibleTextField: View {
    let label: String
    let placeholder: String
    @Binding var text: String
    let hint: String?
    let isSecure: Bool
    let keyboardType: UIKeyboardType

    init(
        _ label: String,
        placeholder: String = "",
        text: Binding<String>,
        hint: String? = nil,
        isSecure: Bool = false,
        keyboardType: UIKeyboardType = .default
    ) {
        self.label = label
        self.placeholder = placeholder
        self._text = text
        self.hint = hint
        self.isSecure = isSecure
        self.keyboardType = keyboardType
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label)
                .font(.caption)
                .foregroundColor(.secondary)
                .accessibilityHidden(true)

            if isSecure {
                SecureField(placeholder, text: $text)
                    .textFieldStyle(.roundedBorder)
                    .keyboardType(keyboardType)
                    .accessibilityLabel(label)
                    .accessibilityHint(hint ?? "")
            } else {
                TextField(placeholder, text: $text)
                    .textFieldStyle(.roundedBorder)
                    .keyboardType(keyboardType)
                    .accessibilityLabel(label)
                    .accessibilityHint(hint ?? "")
            }
        }
    }
}

/// Accessible progress bar
struct AccessibleProgressView: View {
    let value: Double
    let total: Double
    let label: String

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label)
                .font(.caption)
                .accessibilityHidden(true)

            ProgressView(value: value, total: total)
                .accessibilityLabel("\(label): \(Int(value / total * 100)) percent")
                .accessibilityValue("\(Int(value)) of \(Int(total))")
        }
    }
}

// MARK: - Dynamic Type Support

extension Font {
    /// Scale font for accessibility
    static func scaledFont(
        _ font: Font,
        minimumScaleFactor: CGFloat = 0.5,
        maximumScaleFactor: CGFloat = 2.0
    ) -> Font {
        return font
    }
}

/// Text that supports Dynamic Type
struct ScaledText: View {
    let text: String
    let font: Font

    @ScaledMetric private var scaleFactor: CGFloat = 1.0

    var body: some View {
        Text(text)
            .font(font)
    }
}
